File: static/cpp/Reverse_bits.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <random>
#include <vector>

#include "./swap_bits.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::numeric_limits;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

vector<long> precomputed_reverse;

long ReverseX(long x, int n) {
  for (int i = 0, j = n; i < j; ++i, --j) {
    x = SwapBits(x, i, j);
  }
  return x;
}

void CreatePrecomputedTable() {
  for (int i = 0; i < (1 << 16); ++i) {
    precomputed_reverse.emplace_back(ReverseX(i, 15));
  }
}

// @include
long ReverseBits(long x) {
  const int kWordSize = 16;
  const int kBitMask = 0xFFFF;
  return precomputed_reverse[x & kBitMask] << (3 * kWordSize) |
         precomputed_reverse[(x >> kWordSize) & kBitMask] << (2 * kWordSize) |
         precomputed_reverse[(x >> (2 * kWordSize)) & kBitMask] << kWordSize |
         precomputed_reverse[(x >> (3 * kWordSize)) & kBitMask];
}
// @exclude

int main(int argc, char* argv[]) {
  CreatePrecomputedTable();
  if (argc == 2) {
    long x = atoi(argv[1]);
    cout << "sizeof(x) = " << sizeof(x) << endl;
    cout << "x = " << x << ", reverse x = " << ReverseBits(x) << endl;
    cout << ReverseX(x, 63) << endl;
    assert(ReverseBits(x) == ReverseX(x, 63));
  } else {
    default_random_engine gen((random_device())());
    for (int times = 0; times < 1000; ++times) {
      uniform_int_distribution<long> dis(0, numeric_limits<long>::max());
      long x = dis(gen);
      cout << "x = " << x << ", reverse x = " << ReverseBits(x) << endl;
      assert(ReverseBits(x) == ReverseX(x, 63));
    }
  }
  return 0;
}

File: static/python/reverse_bits.py
import sys
import random
import swap_bits


def reverse_x(x, n):
    i, j = 0, n
    while i < j:
        x = swap_bits.swap_bits(x, i, j)
        i += 1
        j -= 1
    return x


PRECOMPUTED_REVERSE = [reverse_x(i, 15) for i in range(1 << 16)]


# @include
def reverse_bits(x):
    MASK_SIZE = 16
    BIT_MASK = 0xFFFF
    return (PRECOMPUTED_REVERSE[x & BIT_MASK] << (3 * MASK_SIZE) |
            PRECOMPUTED_REVERSE[(x >> MASK_SIZE) & BIT_MASK] << (2 * MASK_SIZE) |
            PRECOMPUTED_REVERSE[(x >> (2 * MASK_SIZE)) & BIT_MASK] <<
            MASK_SIZE | PRECOMPUTED_REVERSE[(x >> (3 * MASK_SIZE)) & BIT_MASK])
# @exclude


def main():
    if len(sys.argv) == 2:
        x = int(sys.argv[1])
        print('x = %#x, reverse x = %#x' % (x, reverse_bits(x)))
        print('%#x' % reverse_x(x, 63))
        assert reverse_bits(x) == reverse_x(x, 63)
    else:
        for _ in range(1000):
            x = random.randint(0, sys.maxsize)
            print('x = %#x, reverse x = %#x' % (x, reverse_bits(x)))
            assert reverse_bits(x) == reverse_x(x, 63)


if __name__ == '__main__':
    main()
